### ISSUE TITLE
Make API referentially transparent

### ISSUE THREAD

**Issue by LukaJCB** [OutWatch/outwatch#40]
Kind of related to #28. Currently part of the API as well as part of the internals aren't referentially transparent. We could rewrite the `createHandler` signatures to include some kind of effect managing monad and then use for comprehensions for syntactic sugar. Here's an example:

Current Api:

    def personComponent(labelText: String, texts: Sink[String]) = {
      val firstNames = createStringHandler("")
      val lastNames = createStringHandler("")

      val fullNames = firstNames
        .combineLatestWith(lastNames)((first, last) => s"$first $last")

      div(
        label(labelText),
        input(inputString --> firstNames),
        input(inputString --> lastNames),
        button(click (fullNames) --> texts, "Submit")
      )
    }

Referentially transparent:

    def personComponent(labelText: String, texts: Sink[String]) = for {
      firstNames <- createStringHandler("")
      lastNames <- createStringHandler("")

      fullNames = firstNames
        .combineLatestWith(lastNames)((first, last) => s"$first $last")

      root <- div(
        label(labelText),
        input(inputString --> firstNames),
        input(inputString --> lastNames),
        button(click (fullNames) --> texts, "Submit")
      )

    } yield root
    

The rest of the API could probably stay the same, or atleast with minimal changes. Components without handlers should be completely backwards compatible (albeit not binary compatible). Eager to know what you guys think.

**Comment by mushtaq** [comment]
The existing api looks simpler, what are the advantages of the new one? what is the return type in each case?

**Comment by LukaJCB** [comment]
It is simpler, however it also has some notion of side effects currently. Whenever we call `createHandler` a new Handler gets created that is not equal to when you would call `createHandler` a second time. This breaks referential transparency and can be confusing for some. 

The new proposal would wrap these calls in some kind of Effect type (e.g. Scalaz IO), which would mean building normal VDOM trees would also be wrapped in this type. That means the return type of the `personComponent` as well as of the `div`etc. functions would be something like `VNodeBuilder`. However, I suggest renaming the current `VNode` to something like `VNodeRepresentation` and the suggested `VNodeBuilder` to `VNode`. 
This way, we can get backwards compatibility for components without handlers. :)

**Comment by mushtaq** [comment]
I see that some will prefer effects to be explicitly captured, but I am finding it difficult to imagine a large code base where all calls are wrapped inside for-yield. Is it possible to keep the older api around for those who do not mind the side-effects?

**Comment by LukaJCB** [comment]
Yeah, this is what I'm contemplating currently. Do you have any pain points in particular? We could see if we could address these :)

**Comment by mushtaq** [comment]
I am planning to start ui work on an upcoming project using outwatch so nothing specific to share. My concern is Future types from async calls requiring for-yield themselves shouldn't complicate the code with this change.

**Comment by LukaJCB** [comment]
Wouldn't you use the `HTTP` api for that? In general it might be good to convert Futures to Observables so that OutWatch understands them :)

**Comment by mushtaq** [comment]
Sure, need to explore it further 🙂

**Comment by mariusmuja** [comment]
+1 for a referentially transparent API. I like the idea of having backwards compatibility for components without handlers. I wonder if we go further than that, like having ```VDomModifier``` an effect type as well and moving the for-yield inside ```DomUtils.hyperscriptHelper```?

**Comment by LukaJCB** [comment]
Hey @mariusmuja, sounds like a cool idea, but I'm not really sure what exactly you mean by that. Can you expand? :)

**Comment by mariusmuja** [comment]
@LukaJCB What I meant by that was to have the signature of ```DomUtils.hyperscriptHelper``` like this:

```scala
def hyperscriptHelper(nodeType: String)(args: Eff[VDomModifier]*): Eff[VNode] = {
  ...
}
```
where ```Eff``` is the effect monad. But now I realize that that you already mentioned that VNode components (```div```,...) would be mondad types, so you probably meant that all along.  However, if the EmitterBuilder's ```-->``` and the ReceiverBuilder's ```<--``` return Eff types, that would allow the VDOM building DSL to be stay pretty much the same, wouldn't it?

**Comment by LukaJCB** [comment]
Fixed by #58

**Comment by fdietze** [comment]
I'd like to have some opinions on this alternative implementation:

https://github.com/OutWatch/outwatch/pull/68
